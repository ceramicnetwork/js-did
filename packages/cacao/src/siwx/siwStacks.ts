import { SignatureType, SiwxMessage } from './siwx.js'

export class SiwStacksMessage extends SiwxMessage {
  toMessage(): string {
    return super.toMessage('Stacks')
  }

  signMessage(): string {
    let message: string
    switch (this.type) {
      case SignatureType.PERSONAL_SIGNATURE: {
        message = this.toMessage()
        break
      }

      default: {
        message = this.toMessage()
        break
      }
    }

    return message
  }
}
