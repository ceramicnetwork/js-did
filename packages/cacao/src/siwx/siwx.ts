/* eslint-disable  @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
// @ts-nocheck
import type { Cacao } from '@didtools/codecs'
import { ParsedMessage as ABNFParsedMessage } from '../abnf.js'
import { AccountId, ChainId } from 'caip'

/**
 * Possible message error types.
 */
export enum ErrorTypes {
  /**Thrown when the `validate()` function can verify the message. */
  INVALID_SIGNATURE = 'Invalid signature.',
  /**Thrown when the `expirationTime` is present and in the past. */
  EXPIRED_MESSAGE = 'Expired message.',
  /**Thrown when some required field is missing. */
  MALFORMED_SESSION = 'Malformed session.',
}

/**
 * Possible signature types that this library supports.
 */
export enum SignatureType {
  /**EIP-191 signature scheme */
  PERSONAL_SIGNATURE = 'Personal signature',
}

export class SiwxMessage {
  /**RFC 4501 dns authority that is requesting the signing. */
  domain: string
  /**Ethereum address performing the signing conformant to capitalization
   * encoded checksum specified in EIP-55 where applicable. */
  address: string
  /**Human-readable ASCII assertion that the user will sign, and it must not
   * contain `\n`. */
  statement?: string = undefined
  /**RFC 3986 URI referring to the resource that is the subject of the signing
   *  (as in the __subject__ of a claim). */
  uri: string
  /**Current version of the message. */
  version: string
  /**Randomized token used to prevent replay attacks, at least 8 alphanumeric
   * characters. */
  nonce?: string = undefined
  /**ISO 8601 datetime string of the current time. */
  issuedAt?: string = undefined
  /**ISO 8601 datetime string that, if present, indicates when the signed
   * authentication message is no longer valid. */
  expirationTime?: string = undefined
  /**ISO 8601 datetime string that, if present, indicates when the signed
   * authentication message will become valid. */
  notBefore?: string = undefined
  /**System-specific identifier that may be used to uniquely refer to the
   * sign-in request. */
  requestId?: string = undefined
  /**EIP-155 Chain ID to which the session is bound, and the network where
   * Contract Accounts must be resolved. */
  chainId: string
  /**List of information or references to information the user wishes to have
   * resolved as part of authentication by the relying party. They are
   * expressed as RFC 3986 URIs separated by `\n- `. */
  resources?: Array<string> = undefined
  /**Signature of the message signed by the wallet. */
  signature?: string = undefined
  /**Type of sign message to be generated. */
  type?: SignatureType

  constructor(param: string | Partial<SiwxMessage>) {
    if (typeof param === 'string') {
      const parsedMessage = new ABNFParsedMessage(param)
      this.domain = parsedMessage.domain
      this.address = parsedMessage.address
      this.statement = parsedMessage.statement
      this.uri = parsedMessage.uri
      this.version = parsedMessage.version
      this.nonce = parsedMessage.nonce
      this.issuedAt = parsedMessage.issuedAt
      this.expirationTime = parsedMessage.expirationTime
      this.notBefore = parsedMessage.notBefore
      this.requestId = parsedMessage.requestId
      this.chainId = parsedMessage.chainId
      this.resources = parsedMessage.resources
    } else {
      Object.assign(this, param)
    }
  }

  static fromCacao<T extends SiwxMessage>(this: new (...args: Array<any>) => T, cacao: Cacao): T {
    const account = AccountId.parse(cacao.p.iss.replace('did:pkh:', ''))
    const siwx = new this({
      domain: cacao.p.domain,
      address: account.address,
      uri: cacao.p.aud,
      version: cacao.p.version,
      chainId: new ChainId(account.chainId).reference,
    })

    if (cacao.p.statement) siwx.statement = cacao.p.statement
    if (cacao.p.nonce) siwx.nonce = cacao.p.nonce
    if (cacao.p.iat) siwx.issuedAt = cacao.p.iat
    if (cacao.p.exp) siwx.expirationTime = cacao.p.exp
    if (cacao.p.nbf) siwx.notBefore = cacao.p.nbf
    if (cacao.p.requestId) siwx.requestId = cacao.p.requestId
    if (cacao.p.resources) siwx.resources = cacao.p.resources

    if (cacao.s) {
      if (cacao.s.s) siwx.signature = cacao.s.s
      if (cacao.s.t === 'eip191') siwx.type = SignatureType.PERSONAL_SIGNATURE
    }

    return siwx
  }

  toMessage(chain: string): string {
    return asString(this, chain)
  }
}

export function asLegacyChainIdString(message: SiwxMessage, chainName: string): string {
  const header = `${message.domain} wants you to sign in with your ${chainName} account:`
  const uriField = `URI: ${message.uri}`
  let prefix = [header, message.address].join('\n')
  const versionField = `Version: ${message.version}`

  if (!message.nonce) {
    message.nonce = (Math.random() + 1).toString(36).substring(4)
  }

  const nonceField = `Nonce: ${message.nonce}`

  const suffixArray = [uriField, versionField, nonceField]

  if (message.issuedAt) {
    Date.parse(message.issuedAt)
  }
  message.issuedAt = message.issuedAt ? message.issuedAt : new Date().toISOString()
  suffixArray.push(`Issued At: ${message.issuedAt}`)

  if (message.expirationTime) {
    const expiryField = `Expiration Time: ${message.expirationTime}`

    suffixArray.push(expiryField)
  }

  if (message.notBefore) {
    suffixArray.push(`Not Before: ${message.notBefore}`)
  }

  if (message.requestId) {
    suffixArray.push(`Request ID: ${message.requestId}`)
  }

  if (message.chainId) {
    suffixArray.push(`Chain ID: ${message.chainId}`)
  }

  if (message.resources) {
    suffixArray.push([`Resources:`, ...message.resources.map((x) => `- ${x}`)].join('\n'))
  }

  const suffix = suffixArray.join('\n')

  if (message.statement) {
    prefix = [prefix, message.statement].join('\n\n')
  }

  return [prefix, suffix].join('\n\n')
}

export function asString(message: SiwxMessage, chainName: string): string {
  const header = `${message.domain} wants you to sign in with your ${chainName} account:`
  const uriField = `URI: ${message.uri}`
  let prefix = [header, message.address].join('\n')
  const versionField = `Version: ${message.version}`

  if (!message.nonce) {
    message.nonce = (Math.random() + 1).toString(36).substring(4)
  }

  const nonceField = `Nonce: ${message.nonce}`
  const chainIdField = `Chain ID: ${message.chainId}`

  const suffixArray = [uriField, versionField, chainIdField, nonceField]

  if (message.issuedAt) {
    Date.parse(message.issuedAt)
  }
  message.issuedAt = message.issuedAt ? message.issuedAt : new Date().toISOString()
  suffixArray.push(`Issued At: ${message.issuedAt}`)

  if (message.expirationTime) {
    const expiryField = `Expiration Time: ${message.expirationTime}`

    suffixArray.push(expiryField)
  }

  if (message.notBefore) {
    suffixArray.push(`Not Before: ${message.notBefore}`)
  }

  if (message.requestId) {
    suffixArray.push(`Request ID: ${message.requestId}`)
  }

  if (message.resources && message.resources.length >= 1) {
    suffixArray.push([`Resources:`, ...message.resources.map((x) => `- ${x}`)].join('\n'))
  }

  const suffix = suffixArray.join('\n')

  if (message.statement) {
    prefix = [prefix, message.statement].join('\n\n')
  }

  return [prefix, suffix].join('\n\n')
}
