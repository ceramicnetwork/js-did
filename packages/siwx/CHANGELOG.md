# @didtools/siwx

## 2.0.0

### Major Changes

- Upgrade to multiformats 13 and related packages impacted

## 1.0.0

### Major Changes

- c64ce06: Light-weight parsing of SIWx message: reduces size of bundle
