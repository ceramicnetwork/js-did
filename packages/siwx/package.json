{
  "name": "@didtools/siwx",
  "version": "0.0.1",
  "description": "Typescript library for Sign-In With X",
  "author": "Sergey Ukustov <sergey@ukstv.me>",
  "license": "(Apache-2.0 OR MIT)",
  "type": "module",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    }
  },
  "files": [
    "dist"
  ],
  "engines": {
    "node": ">=14.14"
  },
  "sideEffects": false,
  "scripts": {
    "build:clean": "del dist",
    "build:js": "swc src -d ./dist --config-file ../../.swcrc",
    "build:types": "tsc --emitDeclarationOnly --skipLibCheck",
    "build": "pnpm run build:clean && pnpm run build:types && pnpm run build:js",
    "lint": "eslint src --fix",
    "test": "./node_modules/.bin/tsm ./node_modules/uvu/bin.js . .+\\.test\\.ts",
    "test:watch": "./node_modules/.bin/watchlist --eager src -- pnpm run test",
    "test:ci": "pnpm run test",
    "prepare": "pnpm run build",
    "prepublishOnly": "package-check"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ceramicnetwork/js-did.git"
  },
  "keywords": [
    "DID",
    "identity",
    "OCAP",
    "capabilities",
    "SIWX",
    "signin",
    "authentication"
  ],
  "bugs": {
    "url": "https://github.com/ceramicnetwork/js-did/issues"
  },
  "homepage": "https://github.com/ceramicnetwork/js-did#readme",
  "dependencies": {
    "codeco": "^1.0.1"
  },
  "devDependencies": {
    "@types/node": "^18.15.11",
    "tsm": "^2.3.0",
    "uvu": "^0.5.6",
    "watchlist": "^0.3.1",
    "ts-essentials": "^9.3.1"
  }
}
