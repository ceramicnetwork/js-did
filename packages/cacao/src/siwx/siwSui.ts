import { SignatureType, SiwxMessage } from './siwx.js'


export class SiwSuiMessage extends SiwxMessage {
  toMessage(): string {
    return super.toMessage('Sui')
  }

  signMessage(): Uint8Array {
    let message: Uint8Array
    switch (this.type) {
      case SignatureType.PERSONAL_SIGNATURE: {
        message = new TextEncoder().encode(this.toMessage())
        break
      }

      default: {
        message = new TextEncoder().encode(this.toMessage())
        break
      }
    }

    return message
  }
}