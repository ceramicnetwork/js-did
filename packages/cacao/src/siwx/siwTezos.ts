import { SignatureType, SiwxMessage } from './siwx.js'
import * as uint8arrays from 'uint8arrays'

export class SiwTezosMessage extends SiwxMessage {
  toMessage(): string {
    return super.toMessage('Tezos')
  }

  signMessage(): string {
    let message: string
    switch (this.type) {
      case SignatureType.PERSONAL_SIGNATURE: {
        message = encodePayload(this.toMessage())
        break
      }

      default: {
        message = encodePayload(this.toMessage())
        break
      }
    }

    return message
  }
}

function encodePayload(message: string): string {
  const michelinePrefix = '05'
  const stringPrefix = '01'
  const len = ('0000000' + message.length.toString(16)).slice(-8)

  const text = uint8arrays.toString(uint8arrays.fromString(message, 'utf-8'), 'hex')
  return michelinePrefix + stringPrefix + len + text
}
