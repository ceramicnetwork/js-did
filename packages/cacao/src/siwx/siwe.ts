import { SignatureType, SiwxMessage } from './siwx.js'

export class SiweMessage extends SiwxMessage {
  /**
   * This function can be used to retrieve an EIP-4361 formated message for
   * signature, although you can call it directly it's advised to use
   * [signMessage()] instead which will resolve to the correct method based
   * on the [type] attribute of this object, in case of other formats being
   * implemented.
   * @returns {string} EIP-4361 formated message, ready for EIP-191 signing.
   */
  toMessage(): string {
    return super.toMessage('Ethereum')
  }

  /**
   * This method parses all the fields in the object and creates a sign
   * message according with the type defined.
   * @returns {string} Returns a message ready to be signed according with the
   * type defined in the object.
   */
  signMessage(): string {
    let message: string
    switch (this.type) {
      case SignatureType.PERSONAL_SIGNATURE: {
        message = this.toMessage()
        break
      }

      default: {
        message = this.toMessage()
        break
      }
    }
    return message
  }
}
