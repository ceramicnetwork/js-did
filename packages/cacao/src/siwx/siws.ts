import { SignatureType, SiwxMessage } from './siwx.js'

export class SiwsMessage extends SiwxMessage {
  toMessage(): string {
    return super.toMessage('Solana')
  }

  signMessage(): Uint8Array {
    let message: Uint8Array
    switch (this.type) {
      case SignatureType.PERSONAL_SIGNATURE: {
        message = new TextEncoder().encode(this.toMessage())
        break
      }

      default: {
        message = new TextEncoder().encode(this.toMessage())
        break
      }
    }

    return message
  }
}
