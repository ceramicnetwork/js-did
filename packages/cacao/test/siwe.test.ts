import { SiweMessage } from '../src/siwx/siwe.js'
import { SignatureType } from '../src/siwx/siwx.js'
import { Wallet } from '@ethersproject/wallet'

const SIWE_MESSAGE_PARAMS = {
  domain: 'service.org',
  address: '0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2',
  statement: 'I accept the ServiceOrg Terms of Service: https://service.org/tos',
  uri: 'https://service.org/login',
  version: '1',
  nonce: '32891757',
  issuedAt: '2021-09-30T16:25:24.000Z',
  chainId: '1',
  resources: [
    'ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu',
    'https://example.com/my-web2-claim.json',
  ],
}

describe('Message Generation', () => {
  test('create and validate eip191 string representation', () => {
    const msg = new SiweMessage(SIWE_MESSAGE_PARAMS)
    const asString = msg.toMessage()
    const recovered = new SiweMessage(asString)
    expect(recovered).toEqual(msg)
  })

  test('With optional fields', () => {
    const msg = new SiweMessage(SIWE_MESSAGE_PARAMS)
    expect(msg.toMessage()).toMatchSnapshot()
  })

  test('No optional field', () => {
    const msg = new SiweMessage({
      ...SIWE_MESSAGE_PARAMS,
      resources: [],
    })

    expect(msg.toMessage()).toMatchSnapshot()
  })

  test('Timestamp without microseconds', () => {
    const msg = new SiweMessage({
      ...SIWE_MESSAGE_PARAMS,
      issuedAt: '2021-09-30T16:25:24Z',
    })

    expect(msg.toMessage()).toMatchSnapshot()
  })
})

describe('Ocap', () => {
  const wallet = Wallet.fromMnemonic(
    'despair voyage estate pizza main slice acquire mesh polar short desk lyrics'
  )

  test('Sign message using SiweMessage object', async () => {
    const address = wallet.address
    const fixedDate = new Date('2021-10-14T07:18:41Z')
    const msg = new SiweMessage({
      domain: 'self.id',
      statement: 'Give this app access to your streams',
      nonce: '12345678',
      type: SignatureType.PERSONAL_SIGNATURE,
      address: address,
      chainId: '1',
      issuedAt: fixedDate.toISOString(),
      resources: ['ceramic://bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a'],
      uri: `did:pkh:eip155:1:${address}`,
      version: '0.1',
    })
    expect(msg.toMessage()).toMatchSnapshot()

    const signature = await wallet.signMessage(msg.toMessage())
    expect(signature).toMatchSnapshot()
  })
})
