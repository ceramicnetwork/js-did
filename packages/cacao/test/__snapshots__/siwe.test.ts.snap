// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message Generation No optional field 1`] = `
"service.org wants you to sign in with your Ethereum account:
0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2

I accept the ServiceOrg Terms of Service: https://service.org/tos

URI: https://service.org/login
Version: 1
Chain ID: 1
Nonce: 32891757
Issued At: 2021-09-30T16:25:24.000Z"
`;

exports[`Message Generation Timestamp without microseconds 1`] = `
"service.org wants you to sign in with your Ethereum account:
0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2

I accept the ServiceOrg Terms of Service: https://service.org/tos

URI: https://service.org/login
Version: 1
Chain ID: 1
Nonce: 32891757
Issued At: 2021-09-30T16:25:24Z
Resources:
- ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu
- https://example.com/my-web2-claim.json"
`;

exports[`Message Generation With optional fields 1`] = `
"service.org wants you to sign in with your Ethereum account:
0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2

I accept the ServiceOrg Terms of Service: https://service.org/tos

URI: https://service.org/login
Version: 1
Chain ID: 1
Nonce: 32891757
Issued At: 2021-09-30T16:25:24.000Z
Resources:
- ipfs://Qme7ss3ARVgxv6rXqVPiikMJ8u2NLgmgszg13pYrDKEoiu
- https://example.com/my-web2-claim.json"
`;

exports[`Ocap Sign message using SiweMessage object 1`] = `
"self.id wants you to sign in with your Ethereum account:
0xBd9D9c7DC389715a89fC8149E4a5Be91336B2796

Give this app access to your streams

URI: did:pkh:eip155:1:0xBd9D9c7DC389715a89fC8149E4a5Be91336B2796
Version: 0.1
Chain ID: 1
Nonce: 12345678
Issued At: 2021-10-14T07:18:41.000Z
Resources:
- ceramic://bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a"
`;

exports[`Ocap Sign message using SiweMessage object 2`] = `"0x66e3a8d71a0c15bbec0afe1e6805cb2826375a6d160031364e3d780aecd9fd0630d48860c3a816b51ca108bc8d6545cb8ae900fa9f988cb3fce1de333d31a43d1b"`;
